[cssom] Serialize -webkit-perspective and -webkit-column-break properties

Loirooriol opened the pull request:
#### 3a4c0cbdfa7b088c28fcb57dd3a1c5889af5c58a
<pre>
[cssom] Serialize -webkit-perspective and -webkit-column-break properties
<a href="https://bugs.webkit.org/show_bug.cgi?id=247988">https://bugs.webkit.org/show_bug.cgi?id=247988</a>

Reviewed by Tim Nguyen.

-webkit-perspective can be serialized like perspective. The former also
accepts unitless lengths, but better not serialize to that.

The -webkit-column-break-* properties are serialized in a way analogous
to page-break-*.

Tests: fast/multicol/column-break-serialization.html
       imported/w3c/web-platform-tests/css/cssom/cssom-getPropertyValue-common-checks.html

* LayoutTests/fast/multicol/column-break-serialization-expected.txt: Added.
* LayoutTests/fast/multicol/column-break-serialization.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/cssom/cssom-getPropertyValue-common-checks-expected.txt:
* Source/WebCore/css/StyleProperties.cpp:
(WebCore::StyleProperties::getPropertyValue const):
(WebCore::StyleProperties::breakInsideShorthandValue const):
(WebCore::StyleProperties::pageBreakValue const):
(WebCore::StyleProperties::webkitColumnBreakValue const):
(WebCore::StyleProperties::pageBreakPropertyValue const): Deleted.
* Source/WebCore/css/StyleProperties.h:

Canonical link: <a href="https://commits.webkit.org/257068@main">https://commits.webkit.org/257068@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/239bc5df450e15782e0ae41cc35397f1a0c3a629

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/97729 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/77/builds/6972 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/43/builds/30897 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/8/builds/107220 "Built successfully") | [✅ 🛠 🧪 win](https://ews-build.webkit.org/#/builders/10/builds/167488 "Built successfully and passed tests") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/11/builds/101675 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/76/builds/7384 "Built successfully") | [✅ 🛠 mac-debug](https://ews-build.webkit.org/#/builders/71/builds/35740 "Built successfully") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/36/builds/90115 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/12/builds/103871 "Built successfully") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/103372 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/78/builds/5542 "Passed tests") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/84361 "Built successfully") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/35/builds/32506 "Passed tests") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/9/builds/87416 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/3/builds/89171 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/34/builds/75426 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/81/builds/964 "Built successfully") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/73/builds/20605 "Passed tests") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/82/builds/954 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/70/builds/22100 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/74/builds/4856 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/80/builds/5776 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/60/builds/44561 "Passed tests") | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/79/builds/2209 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/62/builds/41495 "Passed tests") | | 
<!--EWS-Status-Bubble-End-->

nt1m commented on LayoutTests/fast/multicol/column-break-serialization.html:
```diff
@@ -0,0 +1,54 @@
+function* getShorthandsFor(longhand) {
+    yield "-webkit-column-" + longhand;
+    yield "page-" + longhand;
+}
```
Using generators here doesn't seem needed. You could just define variables right before they're used and iterate through them (either nested arrays, Object.entries(), etc.).

Loirooriol replied:
Switched to maps.

nt1m commented on Source/WebCore/css/StyleProperties.cpp:
```diff
@@ -339,6 +339,10 @@ String StyleProperties::getPropertyValue(CSSPropertyID propertyID) const
+    case CSSPropertyWebkitColumnBreakAfter:
+    case CSSPropertyWebkitColumnBreakBefore:
+    case CSSPropertyWebkitColumnBreakInside:
+        return columnBreakPropertyValue(shorthand);
```
nit: I'd name this `webkitColumnBreakValue`.

Loirooriol replied:
Changed to this:
- `breakInsideShorthandValue` for shorthands of `break-inside`
- `pageBreakValue` for `page-break-before` and `page-break-after`
- `webkitColumnBreakValue` for `-webkit-column-break-before` and `-webkit-column-break-after`

nt1m commented on Source/WebCore/css/StyleProperties.cpp:
```diff
@@ -1355,6 +1354,38 @@ String StyleProperties::pageBreakPropertyValue(const StylePropertyShorthand& sho
+    if (!is<CSSPrimitiveValue>(*value))
+        return String();
+
+    CSSValueID valueId = downcast<CSSPrimitiveValue>(*value).valueID();
```
```suggestion
    auto valueID = propertyAsValueID(shorthand.properties()[0]);
    if (!valueID)
        return emptyString();
    // FIXME: Remove this isCSSWideKeyword check after we do this consistently for all shorthands in getPropertyValue.
    if (isCSSWideKeyword(*valueID))
        return nameString(*valueID);
```

This can be shorter with `propertyAsValueID`.

nt1m commented on Source/WebCore/css/StyleProperties.cpp:
```diff
@@ -1355,6 +1354,38 @@ String StyleProperties::pageBreakPropertyValue(const StylePropertyShorthand& sho
+            return value->cssText();
+        return String();
+    case CSSValueAuto:
+        return value->cssText();
```
I'd prefer splitting inside to a separate function, given you can't end up with CSSValueColumn in break-inside.

Also, `emptyString()`/`nameString(*valueID)` would be preferable to `String()`/`cssText()`.

Loirooriol replied:
It can't be `emptyString()` because [`asTextInternal()`](https://searchfox.org/wubkat/rev/fd9594cdf418745718908fd1869a116307a46e7a/Source/WebCore/css/StyleProperties.cpp#1756) checks `isNull()`, and possibly other places too. I can use `emptyString()`, though. But note `String()` is used all over the place.

I could use `nameString(CSSValueAuto)` but this wouldn't use the `cssTextCache`, not sure if that's bad.

nt1m replied:
`cssText()` does seem to use `nameString()` underneath, so I guess either is fine. 

`nullString()` could work instead of `emptyString()`, not sure what's appropriate in your case.

nt1m replied:
Though I guess you won't have access to the value by using `propertyAsValueID()` so, you'll likely need to use `nameString()`. I think bypassing the cache is fine in this case and looking up the string directly in the table is fine. We already do it for font.

Loirooriol committed:
[cssom] Serialize -webkit-perspective and -webkit-column-break properties
https://bugs.webkit.org/show_bug.cgi?id=247988

Reviewed by Tim Nguyen.

-webkit-perspective can be serialized like perspective. The former also
accepts unitless lengths, but better not serialize to that.

The -webkit-column-break-* properties are serialized in a way analogous
to page-break-*.

Tests: fast/multicol/column-break-serialization.html
       imported/w3c/web-platform-tests/css/cssom/cssom-getPropertyValue-common-checks.html

* LayoutTests/fast/multicol/column-break-serialization-expected.txt: Added.
* LayoutTests/fast/multicol/column-break-serialization.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/cssom/cssom-getPropertyValue-common-checks-expected.txt:
* Source/WebCore/css/StyleProperties.cpp:
(WebCore::StyleProperties::getPropertyValue const):
(WebCore::StyleProperties::breakInsideShorthandValue const):
(WebCore::StyleProperties::pageBreakValue const):
(WebCore::StyleProperties::webkitColumnBreakValue const):
(WebCore::StyleProperties::pageBreakPropertyValue const): Deleted.
* Source/WebCore/css/StyleProperties.h:

Canonical link: https://commits.webkit.org/257068@main
